File: Python3/0862_shorest_subarray_with_sum_at_least_k.py
# https://leetcode.cn/problems/shortest-subarray-with-sum-at-least-k/

class Solution:
    def shortestSubarray(self, nums, k):
        """
        Inputs:
            nums: list[int]
            k: int
        Outputs:
            res: int
        """

        preSumArr = [0]
        res = len(nums) + 1
        for num in nums:
            preSumArr.append(preSumArr[-1] + num)

        q = deque()
        for i, curSum in enumerate(preSumArr):
            while q and curSum - preSumArr[q[0]] >= k:
                res = min(res, i - q.popleft())

            while q and preSumArr[q[-1]] >= curSum:
                q.pop()
                
            q.append(i)
        
        if res >= len(nums) + 1:
            return -1
        return res

File: Cpp/0862_shorest_subarray_with_sum_at_least_k.cpp
// https://leetcode.cn/problems/shortest-subarray-with-sum-at-least-k/

class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        int n = nums.size();
        vector<long> preSumArr(n+1, 0);
        for (int i = 0; i < n; ++i) {
            preSumArr[i+1] = preSumArr[i] + nums[i];
        }

        int res = n + 1;
        deque<int> q;
        for (int i = 0; i <= n; ++i) {
            while ((!q.empty()) && (preSumArr[i] - preSumArr[q.front()] >= k)) {
                int tmp = i - q.front();
                q.pop_front();
                if (tmp < res) {
                    res = tmp;
                }
            }
            while ((!q.empty()) && (preSumArr[q.back()] >= preSumArr[i])) {
                q.pop_back();
            }
            q.push_back(i);
        }

        if (res > n) {
            return -1;
        }
        return res;
    }
};
